vendor/github.com/ligato/cn-infra/exec/supervisor/plugin.go

```diff
@@ -0,0 +1,215 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package supervisor
+
+import (
+	"path/filepath"
+	"sync"
+
+	"github.com/ligato/cn-infra/logging"
+
+	"github.com/ligato/cn-infra/exec/processmanager/status"
+
+	pm "github.com/ligato/cn-infra/exec/processmanager"
+	"github.com/ligato/cn-infra/infra"
+	"github.com/pkg/errors"
+)
+
+// Supervisor is a simple interface to gather information about running programs
+type Supervisor interface {
+	// GetProgramNames returns names of all running program instances
+	GetProgramNames() []string
+	// GetProgramByName returns the process instance of a given program
+	GetProgramByName(name string) pm.ProcessInstance
+}
+
+// Plugin is a supervisor plugin representation
+type Plugin struct {
+	// a map of executed programs. Note that the map entries are never deleted,
+	// the entry is still present with last returned state.
+	mx       sync.Mutex
+	programs map[string]*processWithStateChan
+
+	// common channel where all executed programs send their process state
+	hookChan chan *processEvent
+
+	// supervisor configuration
+	config *Config
+
+	Deps
+}
+
+// Deps are supervisor plugin dependencies
+type Deps struct {
+	infra.PluginDeps
+	PM pm.ProcessManager
+}
+
+// helper structure which holds process instance and channel to access its state.
+// Together with name and logger, it is stored as internal plugin cache
+type processWithStateChan struct {
+	process   pm.ProcessInstance
+	stateChan chan status.ProcessStatus
+	svLogger  *SvLogger
+}
+
+// helper structure with program name, status and event type. The object is passed to the hook
+// resolver by every process watcher
+type processEvent struct {
+	name      string
+	state     status.ProcessStatus
+	eventType EventType
+}
+
+// Init supervisor config file, start event watcher and programs
+func (p *Plugin) Init() error {
+	p.Log.SetLevel(logging.DebugLevel)
+
+	// retrieve configuration file
+	if err := p.getConfig(); err != nil {
+		return errors.Errorf("failed to retrieve supervisor config file: %v", err)
+	}
+	if p.config == nil || len(p.config.Programs) == 0 {
+		return errors.Errorf("supervisor config file not defined or does not contain any program")
+	}
+	p.programs = make(map[string]*processWithStateChan)
+	p.hookChan = make(chan *processEvent)
+
+	go p.watchEvents()
+	// start programs in another go routine (do not block init since it may take a while)
+	go p.startPrograms()
+
+	return nil
+}
+
+// Close local resources
+func (p *Plugin) Close() error {
+	for _, program := range p.programs {
+		if err := program.svLogger.Close(); err != nil {
+			p.Log.Errorf("failed to close logger: %v", err)
+		}
+	}
+	close(p.hookChan)
+	return nil
```

Comment:
The `Close` method should stop running programs and wait until all programs have exited.